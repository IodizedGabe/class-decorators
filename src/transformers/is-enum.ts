import { composePropertyDecorators, IsEnumOptions } from '@hippo-oss/dto-decorators';

import { createBasePropertyDecorators } from './base';

export function IsEnum(options: IsEnumOptions): PropertyDecorator {
    return composePropertyDecorators([
        ...createBasePropertyDecorators(options),
    ]);
}
