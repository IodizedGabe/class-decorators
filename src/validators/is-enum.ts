import { composePropertyDecorators, IsEnumOptions } from '@hippo-oss/dto-decorators';
import { IsEnum as IsEnumDecorator } from 'class-validator';

import { createBasePropertyDecorators } from './base';

export function IsEnum(options: IsEnumOptions): PropertyDecorator {
    return composePropertyDecorators([
        ...createBasePropertyDecorators(options),

        // validate data as an enum
        IsEnumDecorator(options.enum, { each: !!options.isArray }),
    ]);
}
