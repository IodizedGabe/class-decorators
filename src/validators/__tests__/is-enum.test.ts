import { validateSync } from 'class-validator';

import { IsEnum } from '../is-enum';

describe('IsEnum', () => {
    enum FixtureEnum {
        Foo = 'Foo',
        Bar = 'Bar',
    }

    it('allows optional enum to be omitted', () => {
        class Fixture {
            @IsEnum({
                enum: FixtureEnum,
                optional: true,
            })
            property!: FixtureEnum;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('allows nullable enum to be omitted', () => {
        class Fixture {
            @IsEnum({
                enum: FixtureEnum,
                nullable: true,
            })
            property!: FixtureEnum;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('does not allow required enum to be omitted', () => {
        class Fixture {
            @IsEnum({
                enum: FixtureEnum,
            })
            property!: FixtureEnum;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('does not allow non enum', () => {
        class Fixture {
            @IsEnum({
                enum: FixtureEnum,
            })
            property!: FixtureEnum;
        }

        const obj = new Fixture();
        obj.property = 'baz' as unknown as FixtureEnum;
        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('allows enum', () => {
        class Fixture {
            @IsEnum({
                enum: FixtureEnum,
            })
            property!: FixtureEnum;
        }

        const obj = new Fixture();
        obj.property = FixtureEnum.Foo;

        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
});
