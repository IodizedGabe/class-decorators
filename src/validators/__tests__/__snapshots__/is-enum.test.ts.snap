// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsEnum does not allow non enum 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isEnum": "property must be a valid enum value",
    },
    "property": "property",
    "target": Fixture {
      "property": "baz",
    },
    "value": "baz",
  },
]
`;

exports[`IsEnum does not allow required enum to be omitted 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "property should not be null or undefined",
      "isEnum": "property must be a valid enum value",
    },
    "property": "property",
    "target": Fixture {},
    "value": undefined,
  },
]
`;
